File: archive/cold-storage/reference/aaes_os_v1/typescript/src/governed_span.ts
import { SpanState, type GovernedSpan, type RuntimeContext } from "./types.js";

let spanCounter = 0;

export function createGovernedSpan(options?: {
  span_id?: string;
  parent_span_id?: string | null;
  runtime_context?: RuntimeContext | null;
}): GovernedSpan {
  const span_id = options?.span_id ?? `span_stub_${++spanCounter}`;
  const parent_span_id = options?.parent_span_id ?? null;
  let state: SpanState = SpanState.INIT;
  let runtime_context = options?.runtime_context ?? null;

  return {
    span_id,
    parent_span_id,
    get state() {
      return state;
    },
    set state(next: SpanState) {
      state = next;
    },
    get runtime_context() {
      return runtime_context;
    },
    set runtime_context(ctx: RuntimeContext | null) {
      runtime_context = ctx;
    },
    close() {
      if (state !== SpanState.RESULTED && state !== SpanState.CLOSED) {
        throw new Error(`AAES_SPAN_STATE_INVALID: cannot close span in state ${state}`);
      }
      state = SpanState.CLOSED;
    },
  };
}

File: archive/cold-storage/reference/aaes_os_v1/rust/src/governed_span.rs
use crate::error::{aaes_err, AaesResult};
use crate::types::{RuntimeContext, SpanState};

#[derive(Debug, Clone)]
pub struct GovernedSpan {
    pub span_id: String,
    pub parent_span_id: Option<String>,
    pub state: SpanState,
    pub runtime_context: Option<RuntimeContext>,
}

impl GovernedSpan {
    pub fn new(span_id: impl Into<String>, parent_span_id: Option<String>) -> Self {
        Self {
            span_id: span_id.into(),
            parent_span_id,
            state: SpanState::Init,
            runtime_context: None,
        }
    }

    pub fn close(&mut self) -> AaesResult<()> {
        if self.state != SpanState::Resulted && self.state != SpanState::Closed {
            return Err(aaes_err(
                "AAES_SPAN_STATE_INVALID",
                format!("cannot close span in state {:?}", self.state),
            ));
        }
        self.state = SpanState::Closed;
        Ok(())
    }
}
